File: group_names_by_char.py
import sys
from collections import Counter


def run(names: list) -> dict:
    dict = {}
    for name in names:
        keys = ''.join(sorted(Counter(name).keys()))
        for key_item in keys:
            if key_item in dict.keys():
                dict[key_item].append(name)
            else:
                dict[key_item] = []
                dict[key_item].append(name)

    for (key, value) in dict.items():
        print(key,':', value)


if __name__ == '__main__':
    list = ['anurag', 'rahul', 'saurabh']
    sys.exit(run(list))

File: GroupNamesByChar.java
import org.locationtech.jts.io.ParseException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class GroupNamesByChar {
    static final int MAX_CHAR = 26;
    static String getKey(List<String> strList)
    {
        String key = "";
        for(String str: strList) {
            str= str.toLowerCase();
            boolean[] visited = new boolean[MAX_CHAR];
            Arrays.fill(visited, false);

            // store all unique characters of current
            // word in key
            for (int j = 0; j < str.length(); j++)
                visited[str.charAt(j) - 'a'] = true;

            for (int j = 0; j < MAX_CHAR; j++)
                if (visited[j])
                    key = key + (char) ('a' + j);
        }
      return key;
    }

    private static  void  findIndex(List<String> names){
        Map<Character,Set<String>> resultMap= new HashMap<>();
        String keys =getKey(names);
        System.out.println(keys);
        for (String name: names){
            char[] charArr= keys.toCharArray();
            for (final char c : charArr) {
                if (name.indexOf(c)!=-1) {
                    resultMap.computeIfAbsent(c, HashSet::new).add(name);
                }
            }
        }
        resultMap.forEach((key, value) -> System.out.println(key + ":" + value));
    }

    public static void main(String[] args) throws ParseException {
        List<String> list= new ArrayList<String>();
        //test data
        list.add("anurag");
        list.add("rahul");
        list.add("Saurabh");
        findIndex(list);

    }

}
